File: Tree/segment tree/python/segment_tree.py
# 307. Range Sum Query - Mutable
# https://leetcode.com/problems/range-sum-query-mutable/description/

class SegmentTree:
    def __init__(self, start, end):
        self.val = 0
        self.start = start
        self.end = end
        self.left = None
        self.right = None

class NumArray(object):
    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.root = self.build_tree(nums, 0, len(nums)-1)

    def build_tree(self, nums, l, r):
        if l > r:
            return None

        # leaf
        if l == r:
            node = SegmentTree(l, r)
            node.val = nums[l]
            return node

        mid = (l + r) / 2
        node = SegmentTree(l, r)
        node.left = self.build_tree(nums, l, mid)
        node.right = self.build_tree(nums, mid + 1, r)
        node.val = node.left.val + node.right.val

        return node

    def update(self, i, val):
        """
        :type i: int
        :type val: int
        :rtype: void
        """
        def change_v(node, i, val):
            #leaf
            if(i == node.start and i == node.end):
                node.val = val
                return
            mid = (node.start + node.end) / 2
            if (i >= mid + 1):
                change_v(node.right, i, val)
            else:
                change_v(node.left, i, val)

            node.val = node.left.val + node.right.val
        change_v(self.root, i, val)


    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        def get_val(node, i, j):
            if node.start == i and node.end == j:
                return node.val

            mid = (node.start + node.end) / 2

            if j <= mid:
                return get_val(node.left, i, j)

            elif i >= mid + 1:
                return get_val(node.right, i, j)
            else:
                return get_val(node.left, i, mid) + get_val(node.right, mid + 1, j)

        return get_val(self.root, i, j)


nums = [1,3,5]
obj = NumArray(nums)
print obj.sumRange(0,2)
obj.update(1,2)
print obj.sumRange(0,2)

File: Tree/segment tree/golang/segmentTree.go
package main

import "fmt"
import "math"

var tree []int
var array []int

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func buildTree(node, start, end int) {
	if start == end {
		tree[node] = array[start]
		return
	}
	mid := (start + end) / 2
	buildTree(2*node, start, mid)
	buildTree(2*node+1, mid+1, end)
	tree[node] = min(tree[2*node], tree[2*node+1])
}

func update(node, start, end, pos, value int) {
	if start > pos || end < pos {
		return
	}
	if start == end {
		tree[node] = value
		return
	}
	mid := (start + end) / 2
	update(2*node, start, mid, pos, value)
	update(2*node+1, mid+1, end, pos, value)
	tree[node] = min(tree[2*node], tree[2*node+1])
}

func query(node, start, end, i, j int) int {
	if start >= i && end <= j {
		return tree[node]
	}
	if start > j || end < i {
		return math.MaxInt32
	}
	mid := (start + end) / 2
	return min(query(2*node, start, mid, i, j), query(2*node+1, mid+1, end, i, j))
}

func main() {
	array = []int{11, 2, 3, -1, 5, 0, 9, 7}
	tree = make([]int, 2*len(array))
	buildTree(1, 0, len(array)-1)

	fmt.Printf("Min [11, 2, 3, -1, 5, 0, 9, 7] = %d\n", query(1, 0, len(array)-1, 0, 7))
	fmt.Printf("Min 11, 2, 3, -1, 5, [0, 9, 7] = %d\n", query(1, 0, len(array)-1, 5, 7))

	update(1, 0, len(array)-1, 3, 4)

	fmt.Printf("Min [11, 2, 3, 4, 5, 0, 9, 7] = %d\n", query(1, 0, len(array)-1, 0, 7))
	fmt.Printf("Min [11, 2, 3, 4, 5], 0, 9, 7 = %d\n", query(1, 0, len(array)-1, 0, 4))
}

